### PR TITLE
Dashboard improvements backend

### PR DESCRIPTION
**Pull request by camsaul**
Backend for most of the stuff from #2326 (I couldn't find the existing PR for this?)

1. Favorite/Unfavorite a Dashboard
    This basically works exactly the same as it does for Cards
2.  Set position for Dashboards 
    Just set it via `PUT /api/dashboard/:id` (This is also basically exactly how it works for DashboardCards)
3.  Archive/unarchive a Dashboard
     Same as with `position`, this is set via the `PUT` endpoint, just like with Cards

### CODE REVIEW

**Comment by tlrobinson** [comment]
Should dashboard ordering be global or per user?

**Comment by camsaul** [comment]
I think global would help new users get started so every time someone signs up they don't have to manually curate the list of important dashboards. People could use favorites if they need to pin something they they need on a regular basis

**Comment by tlrobinson** [comment]
Make sense, but maybe only admins should be able to reorder them?

**Comment by camsaul** [comment]
Ok, Dashboard archiving is live. I'm sure there are still a few rough edges to smooth out but it should be working for the most part. As with the other things this works exactly like Card archiving, e.g. set instead of calling `DELETE /api/dashboard/:id`, you should instead call `PUT /api/dashboard/:id` and set `archived` to `false`.

File: src/metabase/api/dashboard.clj [hunk 1]
```diff
@@ -9,28 +10,43 @@
...
+    (hydrate-favorites <>)))
 
 (defendpoint GET "/"
   "Get `Dashboards`. With filter option `f` (default `all`), restrict results as follows:
 
-  *  `all` - Return all `Dashboards`.
-  *  `mine` - Return `Dashboards` created by the current user."
+  *  `all`      - Return all Dashboards.
+  *  `mine`     - Return Dashboards created by the current user.
+  *  `archived` - Return Dashboards that have been archived. (By default, these are *exculded*.)"
```
Comment (attekei) [inline src/metabase/api/dashboard.clj]: tiny typo, exculded->excluded

File: src/metabase/api/dashboard.clj [hunk 2]
```diff
@@ -88,26 +107,36 @@
...
       (check-superuser)))
-  (check-500 (-> (assoc dashboard :id id)
-                 (dashboard/update-dashboard! *current-user-id*))))
-
-
+  (check-500
+   (db/update! Dashboard id
+     ;; description, position are allowed to be `nil`. Everything else must be non-nil
+     (u/select-keys-when dashboard
+       :present #{:description :position}
```
Comment (attekei) [inline src/metabase/api/dashboard.clj]: Just for making it clear to myself: what are the scenarios when the value of position is `nil`?

File: src/metabase/api/dashboard.clj [hunk 3]
```diff
@@ -88,26 +107,36 @@
...
   [id]
+  (log/warn "DELETE /api/dashboard/:id is deprecated. Instead of deleting a Dashboard, you should change its `archived` value via PUT /api/dashboard/:id.")
   (let [dashboard (write-check Dashboard id)]
-    (db/delete! Dashboard :id id)
+    (db/delete! Dashboard id)
     (events/publish-event! :dashboard-delete (assoc dashboard :actor_id *current-user-id*)))
   generic-204-no-content)
 
 
+;; TODO - param should be `card_id`, not `cardId` (fix here + on frontend at the same time)
```
Comment (attekei) [inline src/metabase/api/dashboard.clj]: will you do the modification to frontend as well?

File: src/metabase/api/metric.clj [hunk 4]
```diff
@@ -49,23 +49,15 @@
 
 (defendpoint PUT "/:id"
   "Update a `Metric` with ID."
-  [id :as {{:keys [name description caveats points_of_interest how_is_this_calculated show_in_getting_started definition revision_message]} :body}]
+  [id :as {{:keys [name description caveats points_of_interest how_is_this_calculated show_in_getting_started definition revision_message], :as body} :body}]
```
Comment (attekei) [inline src/metabase/api/metric.clj]: `{:keys [...]}` here seems obsolete

**Review by attekei — COMMENTED** [review]
api contract looks good, just some random questions/comments

File: src/metabase/api/metric.clj (hunk 4 above)
Comment (camsaul) [inline src/metabase/api/metric.clj]: yup

File: src/metabase/api/dashboard.clj (hunk 3 above)
Comment (camsaul) [inline src/metabase/api/dashboard.clj]: that's a TODO for some other time, no need to tackle it here in this PR

as I'm working on things I add notes like that to go back and clean up later but tackling them would be out of scope for this PR

File: src/metabase/api/dashboard.clj (hunk 2 above)
Comment (camsaul) [inline src/metabase/api/dashboard.clj]: `position` is `nil` by default unless somebody sets it. So if position is `nil` you'll just have to fall back to default sorting for the Dashboards (whatever that is) until somebody starts reördering them

File: src/metabase/api/dashboard.clj (hunk 3 above)
Comment (attekei) [inline src/metabase/api/dashboard.clj]: 👍

File: src/metabase/api/dashboard.clj (hunk 2 above)
Comment (attekei) [inline src/metabase/api/dashboard.clj]: ök!

File: src/metabase/api/dashboard.clj (hunk 1 above)
Comment (camsaul) [inline src/metabase/api/dashboard.clj]: good catch

**Comment by attekei** [comment]
I will merge this so that I can use real API in #4868

**Pull request merged**